api/src/pcapi/core/external_bookings/decorators.py

```diff
@@ -0,0 +1,91 @@
+from typing import Any
+from typing import Callable
+from typing import TypeVar
+
+from pcapi.utils.requests import exceptions as requests_exception
+
+from . import exceptions
+from . import models
+
+
+logger = logging.getLogger(__name__)
+
+
+class ExternalBookingDecoratorException(Exception):
+    pass
+
+
+def _extract_logging_information(client_func: Callable, *args: list[Any], **kwargs: Any) -> dict:
+    """
+    Return a dictionary to be passed to the `extra` parameter of the logging function.
+
+    The returned dictionary contains the following keys:
+        - "cinema_id" (int): The ID of the cinema from the pivot table.
+        - "client" (str): The client used for the request (e.g., `EMSClientAPI`).
+        - "method" (str): The instance method used for the request (e.g., `book_ticket`).
+        - "method_params" (dict): The parameters passed to the method. Example:
+            {
+                "show_id": "12345",             # ID of the show
+                "booking": "<Booking #1>",      # Stringified Booking instance
+                "beneficiary": "<User #3>",     # Stringified User instance
+            }
+    """
+    self = args[0]
+    assert isinstance(self, models.ExternalBookingsClientAPI)  # to make mypy happy
+    # as client_func is an instance method of `ExternalBookingsClientAPI`, self should have a `cinema_id`
+    cinema_id = self.cinema_id
+
+    func_parameters = [param.name for param in inspect.signature(client_func).parameters.values()]
+    method_params = {}
+
+    arg_index = 1  # as the first argument is `self`, start from index 1
+    for arg_value in args[1:]:
+        arg_name = func_parameters[arg_index]
+        method_params[arg_name] = str(arg_value)
+        arg_index += 1
+
+    for k, v in kwargs.items():
+        method_params[k] = str(v)
+
+    return {
+        "client": self.__class__.__name__,  # class name
+        "method": client_func.__name__,  # method name
+        "cinema_id": cinema_id,
+        "method_params": method_params,
+    }
+
+
+def _check_wrapped_function_is_an_instance_method_of_external_booking_client_api(args: list) -> None:
+    if len(args) == 0 or not isinstance(args[0], models.ExternalBookingsClientAPI):
+        raise ExternalBookingDecoratorException(
+            "`catch_request_timeout` can only be applied to an instance method of a class inheriting from `ExternalBookingsClientAPI`"
+        )
+
+
+# typing to ensure that mypy understands the wrapped function signature has not changed
+F = TypeVar("F", bound=Callable[..., Any])
+
+
+def catch_cinema_provider_request_timeout(client_func: F) -> F:
+    """
+    Decorator that catches HTTP Timeout errors, logs them and then raises `ExternalBookingTimeoutException`.
+    This decorator can only be applied to instance methods of classes inheriting from `ExternalBookingsClientAPI`
+    """
+
+    @functools.wraps(client_func)
+    def wrapped_func(*args, **kwargs):  # type: ignore[no-untyped-def]
```

Comment:
Un peu dommage ce `type: ignore`

```suggestion
    def wrapped_func(*args: Any, **kwargs: Any) -> Any:
```

(Ce qui permet de supprimer celui du dessous également)

Reply:
Yes j'ai fait le changement. Par contre je suis obligé de garder : 

```python
return wrapped_func  # type: ignore[return-value]
```
Parce que mypy ne comprend pas que la signature de la fonction wrappé n'a pas changé. Mais le typage est assuré par:

```python
F = TypeVar("F", bound=Callable[..., Any])


def catch_cinema_provider_request_timeout(client_func: F) -> F:
```


Et j'ai vérifié ça marche bien : 
```python
@catch_cinema_provider_request_timeout
def test_func(test_param: int) -> int:
    return test_param * 2


test_func("coucou")
```

Ça raise cette erreur : 
```
src/pcapi/core/external_bookings/decorators.py:95:11: error: Argument 1 to "test_func"
has incompatible type "str"; expected "int"  [arg-type]
    test_func("coucou")
              ^~~~~~~~
Found 1 error in 1 file (checked 1 source file)
```